tokio/tests/sync_mpsc.rs

```diff
@@ -657,3 +663,273 @@ fn recv_timeout_panic() {
+    }
+
+    impl MyActorHandle {
+        pub fn new() -> (Self, MyActor) {
+            let (sender, receiver) = mpsc::channel(8);
+            let actor = MyActor::new(receiver, sender.clone().downgrade());
+
+            (Self { sender }, actor)
+        }
+
+        pub async fn get_unique_id(&self) -> u32 {
+            let (send, recv) = oneshot::channel();
+            let msg = ActorMessage::GetUniqueId { respond_to: send };
+
+            // Ignore send errors. If this send fails, so does the
+            // recv.await below. There's no reason to check the
+            // failure twice.
+            let _ = self.sender.send(msg).await;
+            recv.await.expect("Actor task has been killed")
+        }
+    }
+
+    let (handle, mut actor) = MyActorHandle::new();
+
+    let actor_handle = tokio::spawn(async move { actor.run().await });
+
+    let _ = tokio::spawn(async move {
+        let _ = handle.get_unique_id().await;
+        drop(handle);
+    })
+    .await;
+
+    let _ = actor_handle.await;
+}
+
+static NUM_DROPPED: AtomicUsize = AtomicUsize::new(0);
+
+#[derive(Debug)]
+struct Msg;
+
+impl Drop for Msg {
+    fn drop(&mut self) {
+        NUM_DROPPED.fetch_add(1, Release);
+    }
+}
+
+// Tests that no pending messages are put onto the channel after `Rx` was
+// dropped.
+//
+// Note: After the introduction of `WeakSender`, which internally
+// used `Arc` and doesn't call a drop of the channel after the last strong
+// `Sender` was dropped while more than one `WeakSender` remains, we want to
+// ensure that no messages are kept in the channel, which were sent after
+// the receiver was dropped.
+#[tokio::test(start_paused = true)]
+async fn test_msgs_dropped_on_rx_drop() {
+    fn ms(millis: u64) -> Duration {
+        Duration::from_millis(millis)
+    }
+
+    let (tx, mut rx) = mpsc::channel(3);
+
+    let rx_handle = tokio::spawn(async move {
+        let _ = rx.recv().await.unwrap();
+        let _ = rx.recv().await.unwrap();
+        time::sleep(ms(1)).await;
+        drop(rx);
+
+        time::advance(ms(1)).await;
+    });
+
+    let tx_handle = tokio::spawn(async move {
+        let _ = tx.send(Msg {}).await.unwrap();
+        let _ = tx.send(Msg {}).await.unwrap();
+
+        // This msg will be pending and should be dropped when `rx` is dropped
+        let _ = tx.send(Msg {}).await.unwrap();
+        time::advance(ms(1)).await;
+
+        // This msg will not be put onto `Tx` list anymore, since `Rx` is closed.
+        time::sleep(ms(1)).await;
+        assert!(tx.send(Msg {}).await.is_err());
+
+        // Ensure that third message isn't put onto the channel anymore
+        assert_eq!(NUM_DROPPED.load(Acquire), 4);
+    });
+
+    let (_, _) = join!(rx_handle, tx_handle);
+}
+
+// Tests that a `WeakSender` is upgradeable when other `Sender`s exist.
+#[tokio::test]
+async fn downgrade_upgrade_sender_success() {
+    let (tx, _rx) = mpsc::channel::<i32>(1);
+    let weak_tx = tx.clone().downgrade();
```

Comment:
Now that it takes `&self`, we don't need this `.clone()`. I'm sure the other tests also have extra calls to `.clone()`.
```suggestion
    let weak_tx = tx.downgrade();
```